xpublish_wms/getmap.py

```diff
@@ -0,0 +1,333 @@
+
+    def select_time(self, da: xr.DataArray) -> xr.DataArray:
+        """
+        Ensure time selection
+
+        If time is provided :
+            - use cf_xarray to access time
+            - by default use TIME_CF_NAME
+            - method nearest to ensure at least one result
+
+        Otherwise:
+            - Get latest one
+
+        :param da:
+        :return:
+        """
+        if self.time is not None:
+            da = da.cf.sel(
+                {self.TIME_CF_NAME: self.time},
+                method="nearest"
+            )
+        else:
+            da = da.cf.isel({
+                self.TIME_CF_NAME: -1
+            })
+
+        return da
+
+    def select_custom_dim(self, da: xr.DataArray) -> xr.DataArray:
+        """
+        Select other dimension, ensuring a 2D array
+        :param da:
+        :return:
+        """
+        # Squeeze single value dimensions
+        da = da.squeeze()
+
+        # TODO: Filter dimension from custom query, if any
+
+        # Squeeze multiple values dimensions, by selecting the last value
+        for coord_name in da.cf.coords:
+            if coord_name in ("latitude", "longitude"):
+                continue
+            coord = da.cf.coords[coord_name]
+            if coord.size > 1:
+                da = da.cf.isel({coord_name: -1})
+
+        return da
+
+    def select_grid(self, da: xr.DataArray) -> xr.DataArray:
+        """
+        Select grid and reproject if needed
+        :param da:
+        :return:
+        """
+        # Some basic check for dataset
+        if not da.rio.crs:
+            da = da.rio.write_crs(self.DEFAULT_CRS)
+
+        if self.grid_is_regular(da):
+            da = self.select_regular_grid(da)
+        else:
+            da = self.select_irregular_grid(da)
+        return da
+
+    def grid_is_regular(self, da: xr.DataArray) -> bool:
+        return da.cf.coords['longitude'].dims[0] == da.cf.coords['longitude'].name
+
+    def select_regular_grid(self, da: xr.DataArray) -> xr.DataArray:
+        """
+        Filter regular grid according to WMS request :
+            - bbox
+            - width and height
+        :param da:
+        :return:
+        """
+        # TBC : need to invert lat/lon here ?
+        #minx, miny, maxx, maxy = self.bbox
+        miny, minx, maxy, maxx = self.bbox
+
+        transform = from_bounds(
+            west=minx, south=miny,
+            east=maxx, north=maxy,
+            width=self.width, height=self.height
+        )
+        clipped = da.rio.clip_box(
+            minx=minx, maxx=maxx,
+            miny=miny, maxy=maxy,
+            crs=self.crs
+        )
+        resampled_data = clipped.rio.reproject(
+            dst_crs=self.crs,
+            shape=(self.height, self.width),
+            resampling=Resampling.nearest,
+            transform=transform,
+        )
+
+        return resampled_data
+
+    def select_irregular_grid(self, da: xr.DataArray) -> xr.DataArray:
```

Comment:
Haven't try this code since I don't have any irregular grid dataset

Reply:
Its what im working heaviliy on right now so well just have to check when merging to make sure my work stays with it